// common-lib/exception/index.js
class ValidationError extends Error {
    constructor(args){
        super(`ValidationError missingFields : ${args.missingFields}`);
        this.missingFields = args.missingFields
        this.code = 400
    }
}

class InvalidError extends Error {
    constructor(args){
        super(`InvalidError : ${args}`);
        this.code = 400
    }
}

class ResourceNotFoundError extends Error {
    constructor(args){
        super(`Resource NotFound : ${args}`);
        this.code = 404
    }
}

class NotImplementedError extends Error {
    constructor(args){
        super(`Not Implemented Exception : ${args}`);
        this.code = 500
    }
}

class InternalServerError extends Error {
    constructor(args){
        super(`InternalServerError : ${args}`);
        this.code = 500
    }
}

class UnauthorizedAccessError extends Error {
    constructor(args){
        super(`UnauthorizedAccessError : ${args}`);
        this.code = 403
    }
}

module.exports = {
    ValidationError,
    InvalidError,
    ResourceNotFoundError,
    NotImplementedError,
    InternalServerError,
    UnauthorizedAccessError
}
